### DISCUSSION TITLE
ZodIssueCode type is an interface rather than enum

### DISCUSSION THREAD

**Discussion by kylerush** [colinhacks/zod#1964] [Q&A] [answered]
I really love zod and thank you to everyone for their work on it 🙏 

I am working with zod in GraphQL and I noticed that the `ZodIssueCode` type is an interface rather than an enum. The code [here](https://github.com/colinhacks/zod/blob/master/src/ZodError.ts#L18) suggests that it is intended to be an enum, but the `arrayToEnum` function [here](https://github.com/colinhacks/zod/blob/master/src/helpers/util.ts#L19) seems to return an object rather than an enum.

In the module's build code I see the following in zod/lib/ZodError.dts:

```ts
export declare const ZodIssueCode: {
    invalid_type: "invalid_type";
    invalid_literal: "invalid_literal";
    custom: "custom";
    invalid_union: "invalid_union";
    invalid_union_discriminator: "invalid_union_discriminator";
    invalid_enum_value: "invalid_enum_value";
    unrecognized_keys: "unrecognized_keys";
    invalid_arguments: "invalid_arguments";
    invalid_return_type: "invalid_return_type";
    invalid_date: "invalid_date";
    invalid_string: "invalid_string";
    too_small: "too_small";
    too_big: "too_big";
    invalid_intersection_types: "invalid_intersection_types";
    not_multiple_of: "not_multiple_of";
    not_finite: "not_finite";
};
```

It seems an enum would be a better type. Is there a reason why it is done this way? Would you be open to a pull request changing the ZodIssueCode to an enum?

Thank you!

**Comment by JacobWeisenburger** [comment]
> It seems an enum would be a better type.

Why do you think that? I'm not saying you're wrong, I just want to better understand your reasoning.

**Comment by kylerush** [comment]
Because the type for all intents and purposes is a set of strings, not an object. If you look at the actual outputted type, you see that the object's properties are the same as the values. That redundancy is what makes me think it isn't the best type. The enum type exists for exactly this purpose. It's for a string value that can only be a value in a set of strings.

Additionally, the function's name implies the author new the best type is enum, but I'm confused why the function name is "...toEnum" but it doesn't actually create an enum.

**Comment by JacobWeisenburger** [comment, marked as answer]
I know of two ways of making enums. One way that is native to Typescript, however has a lot of problems and gotchas. The other way is to use a plain object, which is a lot easier to work. To me, this is merely a matter of taste, but I would recommend the plain object way.

```ts
enum Color {
    Red = 'red',
    Green = 'green',
    Blue = 'blue',
}
console.log( Color )
// { Red: 'red', Green: 'green', Blue: 'blue' }

const colorEnum = {
    Red: 'red',
    Green: 'green',
    Blue: 'blue',
}
console.log( colorEnum )
// { Red: 'red', Green: 'green', Blue: 'blue' }
```

Sources:
[JavaScript Enums in 5 Minutes - What they are and how to create them](https://www.youtube.com/watch?v=nPrHbLsqb54)
[TypeScript's Worst Feature (And How To Avoid It)](https://www.youtube.com/watch?v=Anu8vHXsavo)
[Enums considered harmful](https://www.youtube.com/watch?v=jjMbPt_H3RQ)
[How to use TypeScript Enums and why not to, maybe](https://www.youtube.com/watch?v=pWPClHdcvVg)

**Comment by kylerush** [comment]
The sources you linked to are very interesting and help me understand the type implementation in zod better. I think the current implementation doesn't need any change.

Thanks for the learning and your time.

**Reply by JacobWeisenburger** [reply]
If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
